daffodil-core/src/main/scala/org/apache/daffodil/dsom/AnnotatedSchemaComponent.scala

```diff
@@ -200,6 +202,41 @@ abstract class AnnotatedSchemaComponentImpl(
     this(xml, Option(lexicalParent))
 }
 
+/**
+ * Identifies the property environment of a term.
+ *
+ * If two terms have the same propEnv and same def/decl, then some things can be shared
+ * about their implementation.
+ *
+ * If they have different propEnv, then all bets are off.
+ *
+ * Can be used as a key for whether to create new instance in a factory, or whether one
+ * can be re-used because the property environment is the same.
+ *
+ * The trick is that the regular objects which carry properties have location information
+ * unique to where they originate (for diagnostic messaging purposes).
+ *
+ * What we want is pure value based - depends on the properties and their values only.
+ */
+case class PropEnv(localProps: Seq[Set[(String, String)]], defaultPropSource: Seq[Set[(String, String)]], optNext: Option[scala.xml.Node]) {
+  //
+  // We depend on equality on PropEnv being set equality for localProps, pointer equality for defaultPropSource, and optNext.
+}
+
+object PropEnvFactory {
+  private val factory =
+    Memoize1((localProps: Seq[Set[(String, String)]]) =>
```

Comment:
can remove this. I have convinced myself that these are being compared for equality properly.